### PR TITLE
Task2 

### PR DESCRIPTION
**Pull request by VasilMitovClouway**
Task 2 is ready for review. One implementation is missing.

### CODE REVIEW

File: src/com/clouway/adapter/ConnectionProvider.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public class ConnectionProvider implements Provider<Connection> {
+  private final String database;
```
Comment (freekman) [inline src/com/clouway/adapter/ConnectionProvider.java]: you will have to apply the changes form the first PR

File: src/com/clouway/core/PersonRepository.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public interface PersonRepository {
+  void register(Person person);
+
+  void delete(Person person);
```
Comment (freekman) [inline src/com/clouway/core/PersonRepository.java]: .

File: src/com/clouway/core/PersonRepository.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+  Optional<Person> find(String egn);
+
+  List<Person> findAll(String letter);
```
Comment (freekman) [inline src/com/clouway/core/PersonRepository.java]: .

File: src/com/clouway/core/PersonRepository.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+  List<Person> display();
```
Comment (freekman) [inline src/com/clouway/core/PersonRepository.java]: .

**Comment by freekman** [comment]
@VasilMitovClouway  please look at your first PR and apply the changes to your other PR's i dont want to look at the same mistakes again and again

File: src/com/clouway/core/PersonRepository.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public interface PersonRepository {
+  void register(Person person) throws SQLException;
```
Comment (freekman) [inline src/com/clouway/core/PersonRepository.java]: if tomorrow we deside to use some no-sql db we will have to change the imp.
Why are you shoaling me that you are usin Sql DB

File: src/com/clouway/core/PersonRepository.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+  void delete(String egn) throws SQLException;
+
+  Optional<Person> find(String egn) throws SQLException;
+
+  List<Person> findAllStartingWith(String letter) throws SQLException;
+
+  List<Person> display() throws SQLException;
```
Comment (freekman) [inline src/com/clouway/core/PersonRepository.java]: .

File: src/com/clouway/core/PersonRepository.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public interface PersonRepository {
+  void register(Person person) throws SQLException;
+
+  void delete(String egn) throws SQLException;
+
+  Optional<Person> find(String egn) throws SQLException;
+
+  List<Person> findAllStartingWith(String letter) throws SQLException;
```
Comment (freekman) [inline src/com/clouway/core/PersonRepository.java]: findWithMatchingNames is a better suggestion

File: src/com/clouway/core/PersonRepository.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public interface PersonRepository {
+  void register(Person person) throws SQLException;
+
+  void delete(String egn) throws SQLException;
+
+  Optional<Person> find(String egn) throws SQLException;
```
Comment (freekman) [inline src/com/clouway/core/PersonRepository.java]: findById

File: src/com/clouway/core/Provider.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package com.clouway.core;
+
+
+import java.sql.SQLException;
+
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public interface Provider<T> {
+  T get() throws SQLException;
```
Comment (freekman) [inline src/com/clouway/core/Provider.java]: what is this Sql doing here

File: src/com/clouway/adapter/PersistentPersonRepository.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public class PersistentPersonRepository implements PersonRepository {
```
Comment (freekman) [inline src/com/clouway/adapter/PersistentPersonRepository.java]: You don't have a single test for PersistentPersonRepository

File: src/com/clouway/core/TripRepository.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
+package com.clouway.core;
+
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public interface TripRepository {
```
Comment (freekman) [inline src/com/clouway/core/TripRepository.java]: same comments as with People repo

File: src/com/clouway/core/TripRepository.java [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  void register(Trip trip) throws SQLException;
+
+  void delete(String egn) throws SQLException;
```
Comment (freekman) [inline src/com/clouway/core/TripRepository.java]: just a note egn means something in Bulgarian but in  English i am not sure

File: src/com/clouway/adapter/PersistentTripRepository.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+      e.printStackTrace();
+    }
+    finally {
+      try {
+        connection.close();
+      } catch (SQLException e) {
+        e.printStackTrace();
+      }
+    }
+    return null;
```
Comment (freekman) [inline src/com/clouway/adapter/PersistentTripRepository.java]: .

File: src/com/clouway/core/TripRepository.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+  Optional<Trip> find(String egn) throws SQLException;
+
+  List<City> mostVisited() throws SQLException;
```
Comment (freekman) [inline src/com/clouway/core/TripRepository.java]: What is the City doing in the Trip repo

File: src/com/clouway/adapter/PersistentTripRepository.java [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  @Override
+  public List<Trip> display() throws SQLException {
```
Comment (freekman) [inline src/com/clouway/adapter/PersistentTripRepository.java]: .

File: src/com/clouway/buisness/TravelAgency.java [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+import com.clouway.core.City;
+import com.clouway.core.Person;
+import com.clouway.core.Trip;
+
+import java.util.List;
+
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public class TravelAgency {
```
Comment (freekman) [inline src/com/clouway/buisness/TravelAgency.java]: are you sure this class is needed

File: test/com/clouway/buisness/TravelAgencyTest.java [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.*;
+
+
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public class TravelAgencyTest {
```
Comment (freekman) [inline test/com/clouway/buisness/TravelAgencyTest.java]: no no no
You should have separate tests for every PersistentRepo

**Comment by mgenov** [comment]
I though that these tasks are going to be ready till friday. Why they are not ready yet ?

File: test/com/clouway/buisness/PersonRepositoryTest.java [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+    tableManager.deleteTable("TRIP");
+    tableManager.deleteTable("PEOPLE");
+    personRepository.register(person1);
+    personRepository.register(person2);
+    personRepository.register(person3);
+  }
+
+
+  @Test
+  public void registerPerson() throws Exception {
```
Comment (freekman) [inline test/com/clouway/buisness/PersonRepositoryTest.java]: happyPath

File: src/com/clouway/adapter/PersistentPersonRepository.java [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+        e.printStackTrace();
+      }
+    }
+  }
+
+  @Override
+  public Optional<Person> find(String EGN) {
+    Connection connection = provider.get();
+    String query = "SELECT * FROM PEOPLE WHERE EGN=" + EGN;
+    Optional result = Optional.empty();
```
Comment (freekman) [inline src/com/clouway/adapter/PersistentPersonRepository.java]: do you have test for this case

File: test/com/clouway/buisness/PersonRepositoryTest.java [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+    Person expected=new Person("Vasko", "1111111111", 25, "pojoemail@object.com");
+    Optional<Person> actual=personRepository.find("1111111111");
+    assertThat(actual.get(),is(expected));
+  }
+
+  @Test
+  public void findAllStartingWithSequens() throws Exception {
+    List<Person> expected=new LinkedList<>();
+    expected.add(new Person("Vasko", "1111111111", 25, "pojoemail@object.com"));
+    expected.add(new Person("Vaskis", "1111111112", 23, "pojoemail@object.com"));
```
Comment (freekman) [inline test/com/clouway/buisness/PersonRepositoryTest.java]: ass a Person which name does not start with v

File: test/com/clouway/buisness/PersonRepositoryTest.java [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public class PersonRepositoryTest {
+  private ConnectionProvider connectionProvider = new ConnectionProvider("TASK2");
+  private PersistentPersonRepository personRepository = new PersistentPersonRepository(connectionProvider);
+
+  private TableManagerForTest tableManager=new TableManagerForTest(connectionProvider);
+  private CalendarForTest calendar=new CalendarForTest();
+  private Person person1 = new Person("Vasko", "1111111111", 25, "pojoemail@object.com");
```
Comment (freekman) [inline test/com/clouway/buisness/PersonRepositoryTest.java]: are these needed

File: test/com/clouway/buisness/TableManagerForTest.java [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+import com.clouway.core.Provider;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public class TableManagerForTest {
```
Comment (freekman) [inline test/com/clouway/buisness/TableManagerForTest.java]: ?

File: src/com/clouway/core/TripRepository.java [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public interface TripRepository {
+  void register(Trip trip);
+
+  void delete(String egn);
```
Comment (freekman) [inline src/com/clouway/core/TripRepository.java]: is the delete needed in the production

File: test/com/clouway/buisness/TripRepositoryTest.java [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public class TripRepositoryTest {
+  private ConnectionProvider connectionProvider = new ConnectionProvider("TASK2");
+  private PersistentPersonRepository personRepository = new PersistentPersonRepository(connectionProvider);
+  private PersistentTripRepository tripRepository = new PersistentTripRepository(connectionProvider);
+  private TableManagerForTest tableManager = new TableManagerForTest(connectionProvider);
+  private CalendarForTest calendar = new CalendarForTest();
+  private Person person1 = new Person("Vasko", "1111111111", 25, "pojoemail@object.com");
```
Comment (freekman) [inline test/com/clouway/buisness/TripRepositoryTest.java]: .

File: test/com/clouway/buisness/TableManagerForTest.java (hunk 22 above)
Comment (VasilMitovClouway) [inline test/com/clouway/buisness/TableManagerForTest.java]: To delete tables. and display them if i needed it.

File: src/com/clouway/core/TripRepository.java (hunk 23 above)
Comment (VasilMitovClouway) [inline src/com/clouway/core/TripRepository.java]: Not really. Deleted it.

File: test/com/clouway/buisness/TripRepositoryTest.java (hunk 24 above)
Comment (VasilMitovClouway) [inline test/com/clouway/buisness/TripRepositoryTest.java]: After every test i clean the tables. So in order to register a Trip i need to register people first.

File: src/com/clouway/adapter/ConnectionProvider.java (hunk 1 above)
Comment (mgenov) [inline src/com/clouway/adapter/ConnectionProvider.java]: .

File: src/com/clouway/core/PersonRepository.java (hunk 5 above)
Comment (mgenov) [inline src/com/clouway/core/PersonRepository.java]: .

File: src/com/clouway/core/PersonRepository.java (hunk 8 above)
Comment (mgenov) [inline src/com/clouway/core/PersonRepository.java]: It's by EGN. findByEgn(String egn);

File: src/com/clouway/core/Provider.java (hunk 9 above)
Comment (mgenov) [inline src/com/clouway/core/Provider.java]: .

File: src/com/clouway/adapter/ConnectionProvider.java (hunk 1 above)
Comment (VasilMitovClouway) [inline src/com/clouway/adapter/ConnectionProvider.java]: Changed

File: src/com/clouway/core/TripRepository.java (hunk 11 above)
Comment (VasilMitovClouway) [inline src/com/clouway/core/TripRepository.java]: Done

File: src/com/clouway/adapter/PersistentTripRepository.java (hunk 13 above)
Comment (VasilMitovClouway) [inline src/com/clouway/adapter/PersistentTripRepository.java]: Changed.

**Pull request merged**